docs/src/service-workers-guide.md

```diff
@@ -0,0 +1,161 @@
+
+```js
+await expect.poll(() => serviceWorker.evaluate(() => self.registration.active?.state)).toBe('activated');
+```
+
+## Network Events and Routing
+
+A Network Request made within a Service Worker (along with the main script request itself) will have:
+
+* [`event: BrowserContext.request`] event emitted
+* [`method: BrowserContext.route`] will see the request
+* [`method: Request.serviceWorker`] will be set to the [Service Worker] instance, and [`method: Request.frame`] will **throw**
+* [`method: Response.fromServiceWorker`] will return `false`
+
+A Network Request made within a page will have:
+
+* [`event: BrowserContext.request`] event emitted
+* [`event: Page.request`] event emitted
+* [`method: BrowserContext.route`] and [`method: Page.route`] will **not** see the request (if a Service Worker's fetch handler was registered)
+* [`method: Request.serviceWorker`] will be set to `null`, and [`method: Request.frame`] will return the [Frame]
+* [`method: Response.fromServiceWorker`] will return `true` (if a Service Worker's fetch handler was registered)
+
+
+Consider the following Service Worker Main Script:
+
+```js
+// filename: service-worker-main.js
+self.addEventListener("install", function (event) {
+  event.waitUntil(
+    caches.open("v1").then(function (cache) {
+      // 1. Pre-fetches and caches `./addressbook.json`
+      return cache.add("./addressbook.json");
+    })
+  );
+});
+
+// Opt to handle FetchEvent's from the page
+self.addEventListener("fetch", (event) => {
+  event.respondWith(
+    (async () => {
+      // 1. Try to first serve directly from caches
+      let response = await caches.match(event.request);
+      if (response) return response;
+
+      // 2. Re-write request for /foo to /bar
+      if (event.request.url.endsWith("foo")) return fetch("./bar");
+
+      // 3. Prevent `tracker.js` from being retrieved, and returns a placeholder response
+      if (event.request.url.endsWith("tracker.js"))
+        return new Response('conosole.log("no trackers!")', {
+          status: 200,
+          headers: { "Content-Type": "text/javascript" },
+        });
+
+      // 4. Otherwise, fallthrough, perform the fetch and respond
+      return fetch(event.request);
+    })()
+  );
+});
+
+self.addEventListener("activate", (event) => {
+  event.waitUntil(clients.claim());
+});
+```
+
+And a Page that simply registers the Service Worker:
+
+```html
+<!-- filename: index.html -->
+<script>
+  window.registrationPromise = navigator.serviceWorker.register('./service-worker-main.js');
+</script>
+```
+
+On the first visit to the page via [`method: Page.goto`], the following Request/Response events would be emitted:
+
+| Scope   | Target         | URL                       | Routed | fromServiceWorker |
+| -       | -              | -                         | -      | -                 |
+| Context | Frame          | index.html                | Yes    |                   |
```

Comment:
"Scope" is a bit unclear, can we do it like `page.on('request')` vs `context.on('request')`?

Reply:
Agreed, just updated. See screenshot for what it looks like. I also changed the `Scope` heading to `Event` since we then put events underneath it.